lib/fromZodIssue.ts

```diff
@@ -1,120 +1,36 @@
 import * as zod from 'zod';
 
 import {
-  ISSUE_SEPARATOR,
-  PREFIX,
-  PREFIX_SEPARATOR,
-  UNION_SEPARATOR,
-} from './config.ts';
-import { prefixMessage } from './prefixMessage.ts';
-import { joinPath } from './utils/joinPath.ts';
-import { isNonEmptyArray } from './utils/NonEmptyArray.ts';
+  type MessageBuilder,
+  type CreateMessageBuilderProps,
+  type ZodIssue,
+  createMessageBuilder,
+} from './MessageBuilder.ts';
 import { ValidationError } from './ValidationError.ts';
 
-export type ZodIssue = zod.ZodIssue;
-
-export function getMessageFromZodIssue(props: {
-  issue: ZodIssue;
-  issueSeparator: string;
-  unionSeparator: string;
-  includePath: boolean;
-}): string {
-  const { issue, issueSeparator, unionSeparator, includePath } = props;
-
-  if (issue.code === 'invalid_union') {
-    return issue.unionErrors
-      .reduce<string[]>((acc, zodError) => {
-        const newIssues = zodError.issues
-          .map((issue) =>
-            getMessageFromZodIssue({
-              issue,
-              issueSeparator,
-              unionSeparator,
-              includePath,
-            })
-          )
-          .join(issueSeparator);
-
-        if (!acc.includes(newIssues)) {
-          acc.push(newIssues);
-        }
-
-        return acc;
-      }, [])
-      .join(unionSeparator);
-  }
-
-  if (issue.code === 'invalid_arguments') {
-    return [
-      issue.message,
-      ...issue.argumentsError.issues.map((issue) =>
-        getMessageFromZodIssue({
-          issue,
-          issueSeparator,
-          unionSeparator,
-          includePath,
-        })
-      ),
-    ].join(issueSeparator);
-  }
-
-  if (issue.code === 'invalid_return_type') {
-    return [
-      issue.message,
-      ...issue.returnTypeError.issues.map((issue) =>
-        getMessageFromZodIssue({
-          issue,
-          issueSeparator,
-          unionSeparator,
-          includePath,
-        })
-      ),
-    ].join(issueSeparator);
-  }
-
-  if (includePath && isNonEmptyArray(issue.path)) {
-    // handle array indices
-    if (issue.path.length === 1) {
-      const identifier = issue.path[0];
-
-      if (typeof identifier === 'number') {
-        return `${issue.message} at index ${identifier}`;
-      }
+export type FromZodIssueOptions =
+  | {
+      messageBuilder: MessageBuilder;
     }
-
-    return `${issue.message} at "${joinPath(issue.path)}"`;
-  }
-
-  return issue.message;
-}
-
-export type FromZodIssueOptions = {
-  issueSeparator?: string;
-  unionSeparator?: string;
-  prefix?: string | null;
-  prefixSeparator?: string;
-  includePath?: boolean;
-};
+  // maintain backwards compatibility
+  | Omit<CreateMessageBuilderProps, 'maxIssuesInMessage'>;
```

Comment:
[question] Should we go with `Pick` instead of `Omit` in order to be explicit on what we need?

Reply:
I think `Omit` helps us be future-compatible with any changes in `CreateMessageBuilderProps`. The intention is to simply exclude `maxIssuesInMessage`.

Reply:
Yep, I thought of that, I'm just concerned whether we might introduce a new prop in the future that should also be omitted, but we forget to define it in the omit list. I'd prefer explicit `Pick` to be better safe than sorry, but not strong on that at all. Please feel free to resolve if you wish to keep current setup.